src/derivative_utils.jl

```diff
@@ -106,6 +106,17 @@ function calc_J!(nlsolver::NLSolver, integrator, cache::OrdinaryDiffEqMutableCac
   is_compos && (integrator.eigen_est = opnorm(J, Inf))
 end
 
+function calc_J!(nlsolver::NLSolver, integrator, cache::OrdinaryDiffEqConstantCache, is_compos)
+  @unpack t,dt,uprev,u,f,p = integrator
+  if DiffEqBase.has_jac(f)
+    nlsolver.cache.J = f.jac(uprev, p, t)
+  else
+    nlsolver.cache.J = jacobian(nlsolver.uf,uprev,integrator)
```

Comment:
We should update `uf` first, similar to the other implementations of `calc_J!`. This is currently missing in the out-of-place implementations `calc_J`.

Reply:
But why does `calc_J` produce the right answer without updating `uf` ?

Reply:
Since it's updated in `calc_W!` in lines such as https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/dcff0bd6809adc640ce910ddbeeba8ed05091431/src/derivative_utils.jl#L525 (but not for `p`, so that's still not correct). I think we should remove those updates in `calc_W!` since the update should only happen if we actually need `uf`, which is only the case in `calc_J`.